File: app/src/components/RadarDayRows.js
// RADAR 3 / RADAR 4 GÜN SATIRLARI — seçili günün mühürlü değeri + önceki güne
// göre yön oku.
//
// NEDEN AYRI DOSYA: RadarScreen'de bu iki satır çizicisi, kendi yardımcı
// haritalarıyla (gün/kaynak eşlemeleri, önceki-gün arayıcıları) birlikte
// ekranın ortasına dağılmıştı. Bileşenler artık uç yanıtını olduğu gibi alıp
// kendi türetmesini yapıyor; ekran yalnız "hangi gün seçili" ve "hangi satır
// açık" bilgisini veriyor.
//
// İKİ RADAR KARIŞTIRILMAZ:
//  * Radar 4 = gerçek 1/X/2 ORANI (1.61 · 3.20 · 4.25)
//  * Radar 3 = kullanıcıların oynama YÜZDESİ (%62 · %21 · %17)
// Aynı ok/renk dili kullanılır ama birimler farklıdır; metinler bunu her
// yerde ayrıca söyler (bkz. RadarTabHeaders.js).
//
// UYDURMA YOK: değeri olmayan satır boş bırakılmaz, KENDİ sebebini yazar.
// Sebep arka uçta üretilir (notes) — burada üretilmez.
import React from 'react';
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native';
import { colors, spacing, radius } from '../theme';
import { providerLabel, providerColor } from './RadarTabHeaders';
import PlayedDnaPanel from './PlayedDnaPanel';

const fmtOdd = (v) => (v == null ? '—' : Number(v).toFixed(2));

// Yön oku: yükseliş yeşil ▲ · düşüş kırmızı ▼ · sabit sarı =.
// Sabit değer de BİR BİLGİDİR; ok basmamak "veri yok" ile karışırdı.
export function yonOku(cur, prev) {
  if (prev == null || cur == null) return null;
  const d = cur - prev;
  if (Math.abs(d) < 0.005) return { s: '=', c: colors.warning };
  return d > 0 ? { s: '▲', c: colors.success } : { s: '▼', c: colors.danger };
}

// Seçili günden ÖNCEKİ en yakın dolu günü bulur (kıyas için).
// "Bir önceki gün" değil "bir önceki KAYITLI gün": arada boş gün varsa
// kıyas atlanmaz, yoksa hareket görünmez olurdu.
function oncekiDoluGun(days, selIdx, sec) {
  for (let i = selIdx - 1; i >= 0; i -= 1) {
    const v = sec(days[i]);
    if (v) return v;
  }
  return null;
}

function SatirBasi({ item }) {
  return (
    <View style={styles.memTop}>
      <Text style={styles.memNo}>{item.no}</Text>
      <Text style={styles.memTeams} numberOfLines={1}>{item.home} – {item.away}</Text>
    </View>
  );
}

/** 1/X/2 üçlüsü + önceki kayıtlı güne göre yön oku. */
export function OddsTriple({ odds, prev }) {
  return (
    <View style={styles.oddsLine}>
      {[['1', 'home'], ['X', 'draw'], ['2', 'away']].map(([lbl, key]) => {
        const v = odds?.[key];
        const a = yonOku(v, prev?.[key]);
        return (
          <View key={lbl} style={styles.oddsCell}>
            <Text style={styles.oddsK}>{lbl}</Text>
            <Text style={styles.oddsV}>{fmtOdd(v)}</Text>
            {a ? <Text style={[styles.oddsArrow, { color: a.c }]}>{a.s}</Text> : null}
          </View>
        );
      })}
    </View>
  );
}

/**
 * RADAR 4 — bir maçın seçili gündeki mühürlü oranı.
 * @param data /api/radar/daily-odds yanıtı (gün listesi + maç hücreleri)
 */
export function MarketRow({ item, data, day }) {
  const days = data?.days || [];
  const selIdx = days.findIndex((d) => d.date === day);
  const cells = (data?.matches || []).find((m) => m.no === item.no)?.cells || {};
  const notes = (data?.matches || []).find((m) => m.no === item.no)?.notes || {};
  const cell = cells[day] || null;
  // Arka uç sebebi vermiyorsa (eski sürüm) tek jenerik cümleye düşülür.
  const why = cell ? null : (notes[day] || null);
  const prev = cell ? oncekiDoluGun(days, selIdx, (d) => cells?.[d.date]) : null;

  return (
    <View style={styles.memRow}>
      <SatirBasi item={item} />
      {cell ? (
        <View style={styles.oddsWrap}>
          <OddsTriple odds={cell.odds} prev={prev?.odds} />
          <Text style={styles.oddsHint}>{prev ? 'Bir önceki güne göre değişim' : 'İlk kayıtlı gün (kıyas yok)'}</Text>
        </View>
      ) : (
        <View>
          <Text style={styles.memNone}>{why?.text ? `${why.text}.` : 'Bu gün için oran kaydı yok.'}</Text>
          {why?.detail ? <Text style={styles.memWhy}>{why.detail}</Text> : null}
        </View>
      )}
    </View>
  );
}

// Bu hafta veri veren sağlayıcılar SABİT sırayla listelenir; sıra yanıttan
// gelen rastgele diziliş olsaydı aynı maç her yenilemede farklı görünürdü.
const PROVIDER_ORDER = ['nesine', 'misli', 'bilyoner', 'oley'];
export function aktifSaglayicilar(sources) {
  return (sources || []).slice()
    .sort((a, b) => ((PROVIDER_ORDER.indexOf(a) + 1) || 99) - ((PROVIDER_ORDER.indexOf(b) + 1) || 99));
}

/**
 * RADAR 3 — bir maçın seçili gündeki oynanma yüzdeleri, HER KAYNAK AYRI SATIR.
 * Kaynaklar ortalanmaz: iki site farklı yüzde veriyorsa bu bir bilgidir,
 * tek sayıya indirilirse kaybolur.
 * @param data     /api/radar/daily-played yanıtı
 * @param openKey  açık DNA paneli anahtarı ("<maçNo>|<kaynak>")
 */
export function PublicRow({ item, data, day, openKey, onToggleDna, roundId, tick }) {
  const days = data?.days || [];
  const gun = days.find((d) => d.date === day);

  // Seçili gün henüz gelmediyse yüzde BASILMAZ (uydurma yerine dürüst not).
  if (gun?.future) {
    return (
      <View style={styles.memRow}>
        <SatirBasi item={item} />
        <Text style={styles.memNone}>Bu gün henüz gelmedi — yüzde o gün oluştukça dolar.</Text>
      </View>
    );
  }

  const selIdx = days.findIndex((d) => d.date === day);
  const cells = (data?.matches || []).find((m) => m.no === item.no)?.cells || {};
  const bySource = cells[day]?.bySource || null;
  const saglayicilar = aktifSaglayicilar(data?.sources);

  return (
    <View style={styles.memRow}>
      <SatirBasi item={item} />
      {saglayicilar.length ? saglayicilar.map((pv) => {
        const c = bySource?.[pv] || null;
        const prev = c ? oncekiDoluGun(days, selIdx, (d) => cells?.[d.date]?.bySource?.[pv]) : null;
        const key = `${item.no}|${pv}`;
        const acik = openKey === key;
        return (
          <View key={pv}>
            {/* Kaynak satırı: dokununca o kaynağın GEÇMİŞ Oynanma DNA'sı açılır. */}
            <TouchableOpacity
              style={styles.provRow}
              activeOpacity={c ? 0.7 : 1}
              disabled={!c}
              onPress={() => onToggleDna(acik ? null : key)}
            >
              {/* KAYNAK ADI YERİNE RENKLİ NOKTA: sarı=Nesine, turuncu=Misli,
                  yeşil=Bilyoner. Satır kısalır, 15 maç × 3 kaynak okunur kalır.
                  Hangi rengin hangi site olduğu panel başlığındaki "Aktif
                  kaynak" satırında ADIYLA yazılır — kaynak gizlenmez.
                  Erişilebilirlik: renk tek ayırt edici olmasın diye
                  accessibilityLabel kaynağın ADINI söyler. */}
              <View
                style={[styles.provDot, { backgroundColor: providerColor(pv) }]}
                accessibilityLabel={providerLabel(pv)}
                testID={`kaynak-nokta-${pv}`}
              />
              {c ? (
                <View style={styles.provVals}>
                  {['1', 'X', '2'].map((k) => {
                    const v = c.percentages?.[k];
                    const a = yonOku(v, prev?.percentages?.[k]);
                    return (
                      <Text key={k} style={styles.provVal}>
                        {k} %{Math.round(Number(v))}
                        {a ? <Text style={{ color: a.c, fontWeight: '900' }}> {a.s}</Text> : null}
                      </Text>
                    );
                  })}
                  <Text style={styles.provChev}>{acik ? '▾' : '›'}</Text>
                </View>
              ) : <Text style={styles.provNone}>bu gün kayıt yok</Text>}
            </TouchableOpacity>
            {acik ? (
              <PlayedDnaPanel roundId={roundId} no={item.no} source={pv} day={day} tick={tick} />
            ) : null}
          </View>
        );
      }) : <Text style={styles.memNone}>Bu gün için oynanma yüzdesi kaydı yok.</Text>}
    </View>
  );
}

const styles = StyleSheet.create({
  memRow: { backgroundColor: colors.card, borderRadius: radius.md, padding: spacing.md, marginBottom: spacing.sm },
  memTop: { flexDirection: 'row', alignItems: 'center', gap: spacing.md },
  memNo: { color: colors.textMuted, fontSize: 15, fontWeight: '800', width: 22, textAlign: 'center' },
  memTeams: { color: colors.text, fontSize: 14, fontWeight: '800', flex: 1 },
  memNone: { color: colors.textMuted, fontSize: 12, fontStyle: 'italic', marginTop: 6, marginLeft: 34 },
  memWhy: { color: colors.textMuted, fontSize: 11, marginTop: 2, marginLeft: 34, opacity: 0.85 },

  oddsWrap: { marginTop: 8, marginLeft: 34 },
  oddsLine: { flexDirection: 'row', gap: 10, flexWrap: 'wrap' },
  oddsCell: { flexDirection: 'row', alignItems: 'baseline', gap: 4, backgroundColor: colors.surfaceSoft, borderRadius: radius.sm, borderWidth: 1, borderColor: colors.border, paddingHorizontal: 10, paddingVertical: 5 },
  oddsK: { color: colors.textMuted, fontSize: 11, fontWeight: '900' },
  oddsV: { color: colors.text, fontSize: 14, fontWeight: '900' },
  oddsArrow: { fontSize: 11, fontWeight: '900' },
  oddsHint: { color: colors.textMuted, fontSize: 10, fontWeight: '700', marginTop: 5 },

  provRow: { flexDirection: 'row', alignItems: 'center', gap: 8, marginTop: 6, marginLeft: 34, flexWrap: 'wrap' },
  // Kaynak noktası: ad yerine renk. Çerçeve açık zeminde sarıyı görünür tutar.
  provDot: { width: 11, height: 11, borderRadius: 6, borderWidth: 1, borderColor: 'rgba(0,0,0,0.18)' },
  provVals: { flexDirection: 'row', alignItems: 'baseline', gap: 12, flexWrap: 'wrap' },
  provVal: { color: colors.textSoft, fontSize: 13, fontWeight: '800' },
  provNone: { color: colors.textMuted, fontSize: 11, fontStyle: 'italic' },
  provChev: { color: colors.textMuted, fontSize: 13, fontWeight: '900' },
});

File: flutter/lib/features/radar/radar_day_rows.dart
// KAYNAK: app/src/components/RadarDayRows.js — BİREBİR çeviri.
//
// RADAR 3 / RADAR 4 GÜN SATIRLARI — seçili günün mühürlü değeri + önceki güne
// göre yön oku.
//
// İKİ RADAR KARIŞTIRILMAZ:
//  * Radar 4 = gerçek 1/X/2 ORANI (1.61 · 3.20 · 4.25)
//  * Radar 3 = kullanıcıların oynama YÜZDESİ (%62 · %21 · %17)
// Aynı ok/renk dili kullanılır ama birimler farklıdır.
//
// UYDURMA YOK: değeri olmayan satır boş bırakılmaz, KENDİ sebebini yazar.
// Sebep arka uçta üretilir (notes) — burada üretilmez.

import 'package:flutter/material.dart';

import '../../core/theme/tokens.dart';
import 'provider_labels.dart';

String _fmtOdd(Object? v) =>
    v == null ? '—' : (v is num ? v.toStringAsFixed(2) : '$v');

typedef Yon = ({String s, Color c});

/// Yön oku: yükseliş yeşil ▲ · düşüş kırmızı ▼ · sabit sarı =.
/// Sabit değer de BİR BİLGİDİR; ok basmamak "veri yok" ile karışırdı.
Yon? yonOku(Object? cur, Object? prev) {
  if (prev is! num || cur is! num) return null;
  final d = cur - prev;
  if (d.abs() < 0.005) return (s: '=', c: AppColors.warning);
  return d > 0 ? (s: '▲', c: AppColors.success) : (s: '▼', c: AppColors.danger);
}

/// Seçili günden ÖNCEKİ en yakın DOLU günü bulur (kıyas için).
/// "Bir önceki gün" değil "bir önceki KAYITLI gün": arada boş gün varsa kıyas
/// atlanmaz, yoksa hareket görünmez olurdu.
Map? _oncekiDoluGun(List days, int selIdx, Map? Function(Map) sec) {
  for (var i = selIdx - 1; i >= 0; i--) {
    final v = sec(days[i] as Map);
    if (v != null) return v;
  }
  return null;
}

class _SatirBasi extends StatelessWidget {
  const _SatirBasi({required this.item});
  final Map item;

  @override
  Widget build(BuildContext context) => Row(
    children: [
      SizedBox(
        width: 22,
        child: Text(
          '${item['no']}',
          textAlign: TextAlign.center,
          style: const TextStyle(
            color: AppColors.textMuted,
            fontSize: 15,
            fontWeight: AppFont.heavy,
          ),
        ),
      ),
      const SizedBox(width: Spacing.md),
      Expanded(
        child: Text(
          '${item['home']} – ${item['away']}',
          maxLines: 1,
          overflow: TextOverflow.ellipsis,
          style: const TextStyle(
            color: AppColors.text,
            fontSize: 14,
            fontWeight: AppFont.heavy,
          ),
        ),
      ),
    ],
  );
}

/// 1/X/2 üçlüsü + önceki kayıtlı güne göre yön oku.
class OddsTriple extends StatelessWidget {
  const OddsTriple({super.key, required this.odds, this.prev});

  final Map? odds;
  final Map? prev;

  @override
  Widget build(BuildContext context) => Wrap(
    spacing: 10,
    runSpacing: 10,
    children: [
      for (final (lbl, key) in const [
        ('1', 'home'),
        ('X', 'draw'),
        ('2', 'away'),
      ])
        _hucre(lbl, odds?[key], yonOku(odds?[key], prev?[key])),
    ],
  );

  Widget _hucre(String lbl, Object? v, Yon? a) => Container(
    padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 5),
    decoration: BoxDecoration(
      color: AppColors.surfaceSoft,
      borderRadius: AppRadius.smR,
      border: Border.all(color: AppColors.border),
    ),
    child: Row(
      mainAxisSize: MainAxisSize.min,
      crossAxisAlignment: CrossAxisAlignment.baseline,
      textBaseline: TextBaseline.alphabetic,
      children: [
        Text(
          lbl,
          style: const TextStyle(
            color: AppColors.textMuted,
            fontSize: 11,
            fontWeight: AppFont.black,
          ),
        ),
        const SizedBox(width: 4),
        Text(
          _fmtOdd(v),
          style: const TextStyle(
            color: AppColors.text,
            fontSize: 14,
            fontWeight: AppFont.black,
          ),
        ),
        if (a != null) ...[
          const SizedBox(width: 4),
          Text(
            a.s,
            style: TextStyle(
              color: a.c,
              fontSize: 11,
              fontWeight: AppFont.black,
            ),
          ),
        ],
      ],
    ),
  );
}

/// RADAR 4 — bir maçın seçili gündeki mühürlü oranı.
/// [data] `/api/radar/daily-odds` yanıtı (gün listesi + maç hücreleri)
class MarketRow extends StatelessWidget {
  const MarketRow({
    super.key,
    required this.item,
    required this.data,
    required this.day,
  });

  final Map item;
  final Map? data;
  final String? day;

  @override
  Widget build(BuildContext context) {
    final days = (data?['days'] as List?) ?? const [];
    final selIdx = days.indexWhere((d) => (d as Map)['date'] == day);
    final mac = ((data?['matches'] as List?) ?? const [])
        .cast<Map>()
        .where((m) => m['no'] == item['no'])
        .firstOrNull;
    final cells = (mac?['cells'] as Map?) ?? const {};
    final notes = (mac?['notes'] as Map?) ?? const {};
    final cell = cells[day] as Map?;
    // Arka uç sebebi vermiyorsa (eski sürüm) tek jenerik cümleye düşülür.
    final why = cell == null ? notes[day] as Map? : null;
    final prev = cell != null
        ? _oncekiDoluGun(days, selIdx, (d) => cells[d['date']] as Map?)
        : null;

    return _Kutu(
      children: [
        _SatirBasi(item: item),
        if (cell != null)
          Padding(
            padding: const EdgeInsets.only(top: 8, left: 34),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                OddsTriple(
                  odds: cell['odds'] as Map?,
                  prev: prev?['odds'] as Map?,
                ),
                Padding(
                  padding: const EdgeInsets.only(top: 5),
                  child: Text(
                    prev != null
                        ? 'Bir önceki güne göre değişim'
                        : 'İlk kayıtlı gün (kıyas yok)',
                    style: const TextStyle(
                      color: AppColors.textMuted,
                      fontSize: 10,
                      fontWeight: AppFont.bold,
                    ),
                  ),
                ),
              ],
            ),
          )
        else ...[
          _Yok(
            metin: why?['text'] != null
                ? '${why!['text']}.'
                : 'Bu gün için oran kaydı yok.',
          ),
          if (why?['detail'] != null)
            Padding(
              padding: const EdgeInsets.only(top: 2, left: 34),
              child: Opacity(
                opacity: 0.85,
                child: Text(
                  '${why!['detail']}',
                  style: const TextStyle(
                    color: AppColors.textMuted,
                    fontSize: 11,
                  ),
                ),
              ),
            ),
        ],
      ],
    );
  }
}

/// RADAR 3 — bir maçın seçili gündeki oynanma yüzdeleri, HER KAYNAK AYRI SATIR.
///
/// Kaynaklar ORTALANMAZ: iki site farklı yüzde veriyorsa bu bir bilgidir, tek
/// sayıya indirilirse kaybolur.
class PublicRow extends StatelessWidget {
  const PublicRow({
    super.key,
    required this.item,
    required this.data,
    required this.day,
    this.openKey,
    this.onToggleDna,
    this.dnaPanelBuilder,
  });

  final Map item;
  final Map? data;
  final String? day;

  /// Açık DNA paneli anahtarı (`"<maçNo>|<kaynak>"`)
  final String? openKey;
  final ValueChanged<String?>? onToggleDna;

  /// Açılan geçmiş DNA paneli. Verilmezse satır yalnız açılıp kapanır.
  final Widget Function(Object kaynak)? dnaPanelBuilder;

  @override
  Widget build(BuildContext context) {
    final days = (data?['days'] as List?) ?? const [];
    final gun = days.cast<Map>().where((d) => d['date'] == day).firstOrNull;

    // Seçili gün henüz gelmediyse yüzde BASILMAZ (uydurma yerine dürüst not).
    if (gun?['future'] == true) {
      return _Kutu(
        children: [
          _SatirBasi(item: item),
          const _Yok(
            metin: 'Bu gün henüz gelmedi — yüzde o gün oluştukça dolar.',
          ),
        ],
      );
    }

    final selIdx = days.indexWhere((d) => (d as Map)['date'] == day);
    final mac = ((data?['matches'] as List?) ?? const [])
        .cast<Map>()
        .where((m) => m['no'] == item['no'])
        .firstOrNull;
    final cells = (mac?['cells'] as Map?) ?? const {};
    final bySource = (cells[day] as Map?)?['bySource'] as Map?;
    final saglayicilar = aktifSaglayicilar(data?['sources'] as List?);

    return _Kutu(
      children: [
        _SatirBasi(item: item),
        if (saglayicilar.isEmpty)
          const _Yok(metin: 'Bu gün için oynanma yüzdesi kaydı yok.')
        else
          for (final pv in saglayicilar)
            _kaynakSatiri(pv, bySource, cells, days, selIdx),
      ],
    );
  }

  Widget _kaynakSatiri(
    Object pv,
    Map? bySource,
    Map cells,
    List days,
    int selIdx,
  ) {
    final c = bySource?[pv] as Map?;
    final prev = c != null
        ? _oncekiDoluGun(
            days,
            selIdx,
            (d) =>
                ((cells[d['date']] as Map?)?['bySource'] as Map?)?[pv] as Map?,
          )
        : null;
    final key = '${item['no']}|$pv';
    final acik = openKey == key;

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        GestureDetector(
          onTap: c == null ? null : () => onToggleDna?.call(acik ? null : key),
          behavior: HitTestBehavior.opaque,
          child: Padding(
            padding: const EdgeInsets.only(top: 6, left: 34),
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                // KAYNAK RENKLİ NOKTAYLA GÖSTERİLİR. Bahis sitesi adı hiçbir
                // yerde geçmez. Erişilebilirlik: renk tek ayırt edici olmasın
                // diye etiket kaynağın RENK ADINI söyler.
                Semantics(
                  key: Key('kaynak-nokta-${kaynakKodu(pv)}'),
                  label: providerLabel(pv),
                  child: Container(
                    width: 11,
                    height: 11,
                    decoration: BoxDecoration(
                      color: providerColor(pv),
                      shape: BoxShape.circle,
                      border: Border.all(color: const Color(0x2E000000)),
                    ),
                  ),
                ),
                const SizedBox(width: 8),
                if (c == null)
                  const Text(
                    'bu gün kayıt yok',
                    style: TextStyle(
                      color: AppColors.textMuted,
                      fontSize: 11,
                      fontStyle: FontStyle.italic,
                    ),
                  )
                else
                  Expanded(
                    child: Wrap(
                      spacing: 12,
                      runSpacing: 4,
                      crossAxisAlignment: WrapCrossAlignment.center,
                      children: [
                        for (final k in const ['1', 'X', '2'])
                          _yuzde(
                            k,
                            (c['percentages'] as Map?)?[k],
                            yonOku(
                              (c['percentages'] as Map?)?[k],
                              (prev?['percentages'] as Map?)?[k],
                            ),
                          ),
                        Text(
                          acik ? '▾' : '›',
                          style: const TextStyle(
                            color: AppColors.textMuted,
                            fontSize: 13,
                            fontWeight: AppFont.black,
                          ),
                        ),
                      ],
                    ),
                  ),
              ],
            ),
          ),
        ),
        if (acik && dnaPanelBuilder != null) dnaPanelBuilder!(pv),
      ],
    );
  }

  Widget _yuzde(String k, Object? v, Yon? a) => RichText(
    text: TextSpan(
      style: const TextStyle(
        color: AppColors.textSoft,
        fontSize: 13,
        fontWeight: AppFont.heavy,
      ),
      children: [
        TextSpan(text: '$k %${v is num ? v.round() : '—'}'),
        if (a != null)
          TextSpan(
            text: ' ${a.s}',
            style: TextStyle(color: a.c, fontWeight: AppFont.black),
          ),
      ],
    ),
  );
}

class _Kutu extends StatelessWidget {
  const _Kutu({required this.children});
  final List<Widget> children;

  @override
  Widget build(BuildContext context) => Container(
    margin: const EdgeInsets.only(bottom: Spacing.sm),
    padding: const EdgeInsets.all(Spacing.md),
    decoration: BoxDecoration(
      color: AppColors.card,
      borderRadius: AppRadius.mdR,
    ),
    child: Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: children,
    ),
  );
}

class _Yok extends StatelessWidget {
  const _Yok({required this.metin});
  final String metin;

  @override
  Widget build(BuildContext context) => Padding(
    padding: const EdgeInsets.only(top: 6, left: 34),
    child: Text(
      metin,
      style: const TextStyle(
        color: AppColors.textMuted,
        fontSize: 12,
        fontStyle: FontStyle.italic,
      ),
    ),
  );
}
